<a href="https://colab.research.google.com/github/FemiAdesola/Data-Science/blob/main/Sales_Department.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <ins>**UNDERSTAND THE PROBLEM STATEMENT AND BUSINESS CASE**


<table>
  <tr><td>
    <img src="https://drive.google.com/uc?id=1l7bHyrjzq839zVZE06cfdDksLabCN2hg"
         alt="Fashion MNIST sprite"  width="1000">
  </td></tr>
  <tr><td align="center">
    <b>Figure 1. Future Sales Time-series Prediction
  </td></tr>
</table>


# <ins>**SALE DEPARTMENT**
+ For companies to become competitive and skyrocket their growth, they need to leverage Al/ML to develop predictive models to forecast sales in the future.
- Predictive models attempt at forecasting future sales based on historical data while taking into account seasonality effects, demand, holidays, promotions, and competition.
+ In this project, you workas a data scientist in the sales department and the sales team provided you with data from <ins>**1115**</ins> stores.
+  The objective is to predict future daily sales based on the features.


![alt text](https://drive.google.com/uc?id=1eLLebiXwkN6x1dpsopQmkVNkR9zAYL7H)

## **The Key features in sales department**

- **Id**: transactionID (combination of Storeand date)
+ **Store**: unique storeld
+ **Sales**: sales/ day, thisis the target variable
+ **Customers**: number of customers on a given day
+ **Open**: Boolean to saywhethera storeis openor
closed (0 = closed, 1 = open)
+ **Promo**: describes if store is running a promo on thatday or not
+ **StateHoliday**: indicate which state holiday (a = p
ublicholiday, b = Easter holiday, c = Christmas, O
= None)
+ **SchoolHoliday**: indicates if the (Store, Date) was affected by the closure of public schools
+ **StoreType**: categorical variable to indicate type of st ore (a, b, c, d)
+ **Assortment**: a = basic, b = extra, c = extended
+ **CompetitionDistance (meters)**: distance to closest c ompetitor store
+ **CompetitionOpenSince [Month/Year]**: datewhenc ompetition was open
+ **Promo2**: Promo2 is a continuing and consecutive promotion for some stores (0 = storeis not participating, 1 = storeis participating)
+ **Promo2Since [Year/Week]**: date when store started participating in Promo2
+ **Promolnterval**: describes the consecutive intervals Promo2 is started, naming the months promotion is started anew. E.g. "Feb,May,Aug, Nov" meanseach round starts in February, May, August, Novemberof any given year for that store


 Data Source: https://www.kaggle.com/c/rossmann-store-sales/data

#  <ins>**IMPORT LIBRARIES AND DATASET**

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import datetime

In [ ]:
# You will need to mount your drive using the following commands:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


#2.1:  <ins>**IMPORT SALES TRAINING DATA**

In [ ]:
# To include the full link to the csv file containing your dataset
sales_train_df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/train.csv')

<ipython-input-3-a1fa2a7154f5>:2: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  sales_train_df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/train.csv')


In [ ]:
sales_train_df.head(5)
# Data Source: https://www.kaggle.com/c/rossmann-store-sales/data



Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [ ]:
sales_train_df.tail(5)

Store  DayOfWeek        Date  Sales  Customers  Open  Promo  \
1017204   1111          2  2013-01-01      0          0     0      0   
1017205   1112          2  2013-01-01      0          0     0      0   
1017206   1113          2  2013-01-01      0          0     0      0   
1017207   1114          2  2013-01-01      0          0     0      0   
1017208   1115          2  2013-01-01      0          0     0      0   

        StateHoliday  SchoolHoliday  
1017204            a              1  
1017205            a              1  
1017206            a              1  
1017207            a              1  
1017208            a              1

In [ ]:
sales_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB


 + 9 columns in total
+ 8 features, each contains 1017209 data points
+ 1 target variable (sales)

In [ ]:
sales_train_df.describe()

Store     DayOfWeek         Sales     Customers          Open  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
mean   5.584297e+02  3.998341e+00  5.773819e+03  6.331459e+02  8.301067e-01   
std    3.219087e+02  1.997391e+00  3.849926e+03  4.644117e+02  3.755392e-01   
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.800000e+02  2.000000e+00  3.727000e+03  4.050000e+02  1.000000e+00   
50%    5.580000e+02  4.000000e+00  5.744000e+03  6.090000e+02  1.000000e+00   
75%    8.380000e+02  6.000000e+00  7.856000e+03  8.370000e+02  1.000000e+00   
max    1.115000e+03  7.000000e+00  4.155100e+04  7.388000e+03  1.000000e+00   

              Promo  SchoolHoliday  
count  1.017209e+06   1.017209e+06  
mean   3.815145e-01   1.786467e-01  
std    4.857586e-01   3.830564e-01  
min    0.000000e+00   0.000000e+00  
25%    0.000000e+00   0.000000e+00  
50%    0.000000e+00   0.000000e+00  
75%    1.000000e+00   0.000000e+00  
max    1.000000e+00   1.000000e+00

- Average sales amount per day = 5773 Euros, minimum sales per day = 0, maximum sales per day = 41551
+ Average number of customers = 633, minimum number of customers = 0, maximum number of customers = 7388

#2.2:  <ins>**IMPORT STORE INFORMATION DATA**

In [ ]:

# StoreType: categorical variable to indicate type of store (a, b, c, d)
# Assortment: describes an assortment level: a = basic, b = extra, c = extended
# CompetitionDistance (meters): distance to closest competitor store
# CompetitionOpenSince [Month/Year]: provides an estimate of the date when competition was open
# Promo2: Promo2 is a continuing and consecutive promotion for some stores (0 = store is not participating, 1 = store is participating)
# Promo2Since [Year/Week]: date when the store started participating in Promo2
# PromoInterval: describes the consecutive intervals Promo2 is started, naming the months the promotion is started anew. E.g. "Feb,May,Aug,Nov" means each round starts in February, May, August, November of any given year for that store




In [ ]:
# Let's do the same for the store_info_df data
# Note that the previous dataframe includes the transactions recorded per day (in millions)
# This dataframe only includes information about the unique 1115 stores that are part of this study


In [ ]:
# on average, the competition distance is 5404 meters away (5.4 kms)

#  <ins>**EXPLORE DATASET**

#3.1:  <ins>**EXPLORE SALES TRAINING DATA**

In [ ]:
# Let's see if we have any missing data, luckily we don't!


In [ ]:
# Average 600 customers per day, maximum is 4500 (note that we can't see the outlier at 7388!)
# Data is equally distibuted across various Days of the week (~150000 observations x 7 day = ~1.1 million observation)
# Stores are open ~80% of the time
# Data is equally distributed among all stores (no bias)
# Promo #1 was running ~40% of the time
# Average sales around 5000-6000 Euros
# School holidays are around ~18% of the time



In [ ]:
# Let's see how many stores are open and closed!


In [ ]:
# Count the number of stores that are open and closed


In [ ]:
# only keep open stores and remove closed stores


In [ ]:
# Let's drop the open column since it has no meaning now


In [ ]:
# Average sales = 6955 Euros,	average number of customers = 762	(went up)

TASK #3.2: EXPLORE STORES INFORMATION DATA

In [ ]:
# Let's see if we have any missing data in the store information dataframe!


In [ ]:
# Let's take a look at the missing values in the 'CompetitionDistance'
# Only 3 rows are missing


In [ ]:
# Let's take a look at the missing values in the 'CompetitionOpenSinceMonth'
# many rows are missing = 354 (almost one third of the 1115 stores)


In [ ]:
# It seems like if 'promo2' is zero, 'promo2SinceWeek', 'Promo2SinceYear', and 'PromoInterval' information is set to zero
# There are 354 rows where 'CompetitionOpenSinceYear' and 'CompetitionOpenSinceMonth' is missing
# Let's set these values to zeros


In [ ]:
# There are 3 rows with 'competitionDistance' values missing, let's fill them up with with average values of the 'CompetitionDistance' column


In [ ]:
# half of stores are involved in promo 2
# half of the stores have their competition at a distance of 0-3000m (3 kms away)

#3.3:  <ins>**EXPLORE MERGED DATASET**

In [ ]:
# Let's merge both data frames together based on 'store'


In [ ]:

# customers and promo are positively correlated with the sales
# Promo2 does not seem to be effective at all

In [ ]:

# Customers/Prmo2 and sales are strongly correlated

In [ ]:
# Let's separate the year and put it into a separate column


In [ ]:
# Let's do the same for the Day and Month


In [ ]:
# Let's take a look at the average sales and number of customers per month
# 'groupby' works great by grouping all the data that share the same month column, then obtain the mean of the sales column
# It looks like sales and number of customers peak around christmas timeframe



In [ ]:
# Let's take a look at the sales and customers per day of the month instead
# Minimum number of customers are generally around the 24th of the month
# Most customers and sales are around 30th and 1st of the month



In [ ]:
# Let's do the same for the day of the week  (note that 7 = Sunday)



#  <ins>**UNDERSTAND THE INTUITION BEHIND FACEBOOK PROPHET**

## <ins>**FACEBOOK PROPHET**
+ Prophet is open source software released by Facebook's Core Data Science team.
+ Prophet is a procedure for forecasting time series data based on an additive model where non-linear trends are fit with yearly, weekly, and daily seasonality, plus holiday effects.
+ Prophet works best with time series that have strong seasonal effects and several seasons of historical data.
+ For more information, please check this out:
    + https://research.fb.com/prophet-forecasting-at-scale/
    + https://facebook.github.io/prophet/docs/quickstart.html#python-api
  
### <ins>**ACCURATE AND FAST**
+ Facebook teams uses Prophet for accurate forecasting and planning.
+ Prophet can generate results in seconds.

### <ins>**AUTOMATIC**
+ No need to perform data preprocessing.
+ Prophet works with missing data with several outliers.

### <ins>**DOMAIN KNOWLEDGE INTEGRATION**
+ Users can tweak forecast by manually adding domain
specific knowledge.

![alt text](https://drive.google.com/uc?id=1CZ24f-TbnRzaXV9Arke0fNTUm7Kon1gK)

# <ins>**TRAIN THE MODEL PART A**

In [ ]:
# import prophet

#  <ins>**TRAIN THE MODEL PART B**


   - StateHoliday: indicates a state holiday. Normally all stores, with few exceptions, are closed on state holidays. Note that all schools are closed on public holidays and weekends. a = public holiday, b = Easter holiday, c = Christmas, 0 = None
   - SchoolHoliday: indicates if the (Store, Date) was affected by the closure of public schools
  






In [ ]:
# Get all the dates pertaining to school holidays


In [ ]:
# Get all the dates pertaining to state holidays


In [ ]:
# concatenate both school and state holidays


In [ ]:
# Let's make predictions using holidays for a specific store
